Add API documentation (#373)
YOURLS/website: docs/guide/advanced/api.md

# API

YOURLS comes with an API wrapper, `yourls-api.php` at the root of your install.
The API allows to:

- Generate or get existing short URLs, with sequential or custom keyword.
- Get some statistics about your links: top clicked links, least clicked links, newest links.
- Return your preferred output format: JSON, XML, or simple raw text.
- Authenticate either with login/password or using a secure [passwordless](passwordless-api) mechanism.
- Even define your own custom API requests and actions.

## Usage

> Parameters must be sent to `https://your-own-domain-here.com/yourls-api.php` either via GET or POST. (Remember to URL-encode parameters if via GET!)

Reviewer: ```suggestion
Parameters must be sent to `https://your-own-domain-here.com/yourls-api.php` either via GET or POST.

:::tip
Remember to URL-encode parameters if via GET
:::
```

# API

YOURLS comes with an API wrapper, `yourls-api.php` at the root of your install.
The API allows to:

- Generate or get existing short URLs, with sequential or custom keyword.
- Get some statistics about your links: top clicked links, least clicked links, newest links.

> - Return various output format: JSON, XML, or simple raw text.

Reviewer 2: ```suggestion
- Return your preferred output format: JSON, XML, or simple raw text.
```

- Authenticate either with login/password or using a secure [passwordless](passwordless-api) mechanism.
- Even define your own custom API requests and actions.

## Usage

> Parameters has to be sent to `http://your-own-domain-here.com/yourls-api.php` either via GET or POST (remember to URL-encode parameters if via GET).

Reviewer 2: ```suggestion
Parameters must be sent to `http://your-own-domain-here.com/yourls-api.php` either via GET or POST. (Remember to URL-encode parameters if via GET!)
```

Does the website support "admonitions" like GH's extended Markdown? The parenthetical could be an "important" aside, like below.

----

Parameters must be sent to `http://your-own-domain-here.com/yourls-api.php` either via GET or POST.

> [!IMPORTANT]
> Remember to URL-encode parameters if using the GET request method.

> These parameters are:

Reviewer 2: ```suggestion

The built-in API parameters are:
```

    - the `filter`: either `"top"`, `"bottom"` , `"rand"`, or `"last"`
    - the `limit` (maximum number of links to return)
    - `output` format: either `"jsonp"`, `"json"`, or `"xml"`

  - `action = "db-stats"`: get global link and click count.
    - `output` format: either `"jsonp"`, `"json"`, or `"xml"`

  - `action = "version"`: get YOURLS version.
    - `output` format: either `"jsonp"`, `"json"`, `"xml"`, or `"simple"`

## Examples

### `GET` request with JavaScript (using jQuery) to shorten a URL

#### JavaScript Request

```javascript
var api_url = 'http://sho.rt/yourls-api.php'
var response = $.get(
  api_url,
  {
    username: 'your_username',
    password: 'your_password',
    action: 'shorturl',
    format: 'json',
    url: 'http://ozh.org/',
  },
  // callback function that will deal with the server response
  function (data) {
    // now do something with the data, for instance show new short URL
    alert(data.shorturl)
  },
)
```

#### JSON Return

```json
{
  "url": {
    "keyword": "ozh",
    "url": "http:\/\/ozh.org",
    "title": "Ozh RICHARD \u00ab ozh.org",
    "date": "2014-10-24 16:01:39",
    "ip": "127.0.0.1"
  },
  "status": "success",

>   "message": "http:\/\/ozh.org added to database",

Reviewer 2: ```suggestion
  "message": "https:\/\/ozh.org added to database",
```

  "title": "Ozh RICHARD \u00ab ozh.org",

>   "shorturl": "http:\/\/sho.rt\/1f",

Reviewer 2: ```suggestion
  "shorturl": "https:\/\/sho.rt\/1f",
```

  "statusCode": "200"
}
```

### `POST` request with PHP to expand a short URL.

#### PHP Request

```php
<?php
$username = 'your_username';
$password = 'your_password';

> $api_url =  'http://sho.rt/yourls-api.php';

Reviewer 2: ```suggestion
$api_url =  'https://sho.rt/yourls-api.php';
```

// Init the CURL session
$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, $api_url);
curl_setopt($ch, CURLOPT_HEADER, 0);            // No header in the result
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true); // Return, do not echo result
curl_setopt($ch, CURLOPT_POST, 1);              // This is a POST request
curl_setopt($ch, CURLOPT_POSTFIELDS, array(     // Data to POST
    'shorturl' => 'ozh',
    'format'   => 'json',
    'action'   => 'expand',
    'username' => $username,
    'password' => $password
));

// Fetch and return content
$data = curl_exec($ch);
curl_close($ch);

// Do something with the result. Here, we echo the long URL
$data = json_decode( $data );
echo $data->longurl;
```

#### XML Return

```xml
<result>
    <keyword>ozh</keyword>
    <shorturl>http://sho.rt/ozh</shorturl>

>     <longurl>http://ozh.org/</longurl>

Reviewer 2: ```suggestion
    <shorturl>https://sho.rt/ozh</shorturl>
    <longurl>https://ozh.org/</longurl>
```
